### PR TITLE
Move SPI bit writes to the right clock phase.

### LINKED ISSUE #20: Bitbang SPI phase=1 lowest bit is wrong
### CircuitPython version

```python
Adafruit CircuitPython 6.1.0 on 2021-01-21; Adafruit QT Py M0 with samd21e18 also
Adafruit CircuitPython 7.0.0-beta.0-161-g65753a1c2 on 2021-09-01; with samd51j19
```

### Code/REPL

```python
# 09/03/2021
# looses both least significant bits
# tested on QT Py and Metro_M4_Express

import board
import digitalio
import adafruit_bitbangio as bitbangio


nEN = digitalio.DigitalInOut(board.A1)    
nEN.direction = digitalio.Direction.OUTPUT

spi = bitbangio.SPI(board.SCK, MOSI=board.MOSI, MISO=board.MISO)

def TestSPI():  
    select = nEN
    select.value = True
    while not spi.try_lock():
        pass
    spi.configure(baudrate=100000, polarity=0, phase=1, bits=8 )
    bytes_read = bytearray(4)
    select.value = False
    bytes_read = bytearray([0x33])  # changes 0x33 to 0x32 
    spi.write(bytes_read)
    bytes_read = bytearray([0x0b])  # changes 0x0b to 0x0a
    spi.write(bytes_read)
    select.value = True
    spi.unlock()
```

### Behavior

image

### Description

_No response_

### Additional information

This works correctly:
from adafruit_bus_device.spi_device import SPIDevice

### PR DESCRIPTION
**Pull request by kbsriram**
- Updated all SPI methods that write bits, so they clock the data on the correct clock transition.
- Added tests with a small logic net simulation to verify bits are read and written correctly.

Technically I think the underlying issue could be resolved without this fix, but I also noticed some issues with the current implementation noted in https://github.com/adafruit/Adafruit_CircuitPython_BitbangIO/issues/20#issuecomment-1964965587

The summary is that all the SPI methods that write bits clock it out on the incorrect clock transition. These examples use the following snippet:
```
 with adafruit_bitbangio.SPI(clock=self.clock_pin, MOSI=self.copi_pin) as spi:
        spi.try_lock()
        spi.configure(baudrate=100, polarity=polarity, phase=phase, bits=8)
        self._enable_net(1)
        spi.write(b"\x55")
        self._enable_net(0)
```
Currently this produces this waveform for `polarity=0` `phase=0` :
write_issue_0_0
The controller clocks out the bit on the rising transition when it's also being read.  However the code likely also works most of the time as the sequence in which the pins are updated in the code are just enough to change the data line right before the clock transitions.

With this PR, the waveform now clocks out the bits at the right transition, and hopefully is more reliable.
fixed_0_0

Unfortunately the same situation exists in all modes, and all the write/read-write methods. This is an example when `polarity=0` and `phase=1` for instance, and data `0x55` and `0xaa`:
write_issue_0_1
write_issue_0_1_1

With this PR, the waveform now shows this for `polarity=0` and `phase=1` for data `0x55` and `0xaa`:
fixed_0_1
fixed_0_1_1

(This PR is large-ish though mostly to do with a test setup that seemed useful to sanity check the fiddly state changes, also in the hope it may be handy to check the I2C code later.)

Resolves https://github.com/adafruit/Adafruit_CircuitPython_BitbangIO/issues/20

### CODE REVIEW

File: tests/simulator.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+                    vcdfile.write(f"#{change.time_us}\n")
+                    last_us = change.time_us
+                vcdfile.write(f"{_level_to_vcd(change.level)}{change.net_name}\n")
+
+
+# module global/singleton
+engine = Engine()
+
+
+class FakePin:
```
Comment (tannewt) [inline tests/simulator.py:116]: I'd call it a DigitalInOut. We actually have a Pin class that does nothing currently. We've discussed enhancing it and calling it Pad though.

```suggestion
class FakeDigitalInOut:
```

**Review by tannewt — CHANGES_REQUESTED** [review]
Very cool! One naming request. Mind checking the C implementation of bitbangio too? https://github.com/adafruit/circuitpython/blob/main/shared-module/bitbangio/SPI.c I think this code was based on it.

**Comment by kbsriram** [comment]
Thanks for reviewing @tannewt !

Re: `Pin` -> `DigitalInOut` - just wanted to double-check, as the current library code is [expecting a `Pin`](https://github.com/adafruit/Adafruit_CircuitPython_BitbangIO/blob/2410e85a2faaa0c408aa6bb91b5817617cf23540/adafruit_bitbangio.py#L311) in the constructor and internally [creates another `DigitalInOut`](https://github.com/adafruit/Adafruit_CircuitPython_BitbangIO/blob/2410e85a2faaa0c408aa6bb91b5817617cf23540/adafruit_bitbangio.py#L330) from it. So I ended up faking the `Pin` class implementation so I could pass it into the constructor, and letting it use the real `DigitalInOut` (which in tests/CPython would come from blinka of course...) Let me know if you had a different idea in mind? Maybe an alternative is when the API changes to passing in a `DigitalInOut` instead of a `Pin`, we could update the test fake then?

Re: comparing [bitbangio/SPI.c](https://github.com/adafruit/circuitpython/blob/main/shared-module/bitbangio/SPI.c) (just for context you're probably aware the native and python implementations started to diverge in terms of how the bits are clocked out in https://github.com/adafruit/Adafruit_CircuitPython_BitbangIO/pull/30)

That said, I think there are some corner cases that appear to need updates in the native code as well, as far as I could tell. For example, with this test function and monitoring the pins:

```python
import board
import bitbangio
import digitalio
import time


def doit():
    time.sleep(3)
    with digitalio.DigitalInOut(board.GP13) as en:
        en.switch_to_output(value=True)
        time.sleep(1)
        with bitbangio.SPI(clock=board.GP15, MOSI=board.GP14) as spi:
            spi.try_lock()
            spi.configure(baudrate=2, polarity=1, phase=0, bits=8)
            en.switch_to_output(value=False)
            spi.write(b"\xaa")
            en.switch_to_output(value=True)
```
I see this output:
test_1_0_0xaa_full

I think the code implicitly assumes the clock is always in the idle state when the bytes start [getting clocked out](https://github.com/adafruit/circuitpython/blob/73fc021c1ac9f40669643db5217c9a0e91840ebf/shared-module/bitbangio/SPI.c#L154) but for `polarity=1` it's not typically the case - and in this situation it ends up essentially missing a clock transition I think, and could cause issues for the first bit. E.g. sigrok correctly identifies the transaction as starting on the enable, as the clock is active at that point. But it doesn't decode the bits.

(The fixed code in this PR for a similar test gives this)
fix_1_0_0xaa

**Comment by tannewt** [comment]
> Re: `Pin` -> `DigitalInOut` - just wanted to double-check, as the current library code is [expecting a `Pin`](https://github.com/adafruit/Adafruit_CircuitPython_BitbangIO/blob/2410e85a2faaa0c408aa6bb91b5817617cf23540/adafruit_bitbangio.py#L311) in the constructor and internally [creates another `DigitalInOut`](https://github.com/adafruit/Adafruit_CircuitPython_BitbangIO/blob/2410e85a2faaa0c408aa6bb91b5817617cf23540/adafruit_bitbangio.py#L330) from it. So I ended up faking the `Pin` class implementation so I could pass it into the constructor, and letting it use the real `DigitalInOut` (which in tests/CPython would come from blinka of course...) Let me know if you had a different idea in mind? Maybe an alternative is when the API changes to passing in a `DigitalInOut` instead of a `Pin`, we could update the test fake then?

Ah, right. Let's leave as-is for now then. We'll want to move both over when we do.

**Review by tannewt — APPROVED** [review]
Thank you!
